osidb/mixins.py

```diff
@@ -657,16 +657,47 @@ def alert(
                 uuid.UUID(acl) for acl in generate_acls([settings.PUBLIC_WRITE_GROUP])
             ]
 
-        Alert.objects.create_alert(
-            name=name,
-            object_id=self.uuid,
-            content_type=ContentType.objects.get_for_model(self),
-            description=description,
-            alert_type=alert_type,
-            resolution_steps=resolution_steps,
-            acl_read=acl_read,
-            acl_write=acl_write,
-        ).save()
+        # Alert.objects.create_alert().save() can raise IntegrityError
+        # (duplicate key value violates unique constraint "unique Alert for name and object").
+        #
+        # The database internally already uses a lock to ensure the uniqueness even when
+        # multiple threads each try to create the same row concurrently.
+        #
+        # Therefore, the database handles parallelism for us already.
+        #
+        # If this raises IntegrityError, it means that this was one of multiple
+        # threads trying to create the same Alert at the same time (the row
+        # not existing at the start of the execution of create_alert() for all
+        # threads, but already existing for one of the threads at the point of
+        # executing save()).
+        #
+        # Therefore, the Alert created by the other thread is very recent.
+        #
+        # Since all Alerts of the same name have practically the same meaning,
+        # it is fine to give up in this thread without retrying to rewrite the
+        # just-created Alert by practically the same information.
+        #
+        # Not using select_for_update() because we don't care if an alert (row) is
+        # overwritten, because all such versions have the same meaning.
+        #
+        # References:
+        # - https://www.postgresql.org/docs/16/index-unique-checks.html#:~:text=At%20present%2C%20only-,b%2Dtree,-supports%20it.)%20Columns
+        # - https://www.postgresql.org/docs/16/locking-indexes.html#:~:text=B%2Dtree%2C,without%20deadlock%20conditions.
```

Comment:
We use postgresql version 13, not 16.

Reply:
Just locally, stage/prod/uat are on aurora anyway :smile:

Reply:
It explains core concepts that are practically timeless anyway. This hasn't changed for 22 years, which is the furthest into the past the relevant documentation can be found quickly. ([1](https://www.postgresql.org/docs/7.2/indexes-unique.html#:~:text=only%20B%2Dtree%20indexes%20can%20be%20declared%20unique), [2](https://www.postgresql.org/docs/7.2/locking-indexes.html#:~:text=B%2Dtree%20indexes,for%20concurrent%20applications.))